### PR TITLE
Improve SYCL reduction performance: RangePolicy

### PR DESCRIPTION
**Pull request by masterleinad**
Part of https://github.com/kokkos/kokkos/pull/6035.
This limits the number of workgroups, effectively processing multiple work items per thread, and increases the maximum workgroup size for Intel GPUs. Technically, `sycl::info::kernel_device_specific::work_group_size` should be the maximum usable value for the workgroup size but it turns out I could still choose `sycl::info::device::max_work_group_size` and got better performance.

Also, fix matching `use_shuffle_based_algorithm`(is the reference not a pointer) with `ReducerType::static_value_size()  ` (0 for array reductions).

### CODE REVIEW

**Comment by masterleinad** [comment]
Only `HIP-ROCm-5.2-C++20` is timing out.

**Review by nliber — APPROVED** [review]

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp [hunk 1]
```diff
@@ -404,42 +403,60 @@ class ParallelReduce<CombinedFunctorReducerType, Kokkos::RangePolicy<Traits...>,
...
-        const size_t wgroup_size =
-            static_cast<size_t>(max / multiple) * multiple;
+        auto max_local_memory =
+            q.get_device().get_info<sycl::info::device::local_mem_size>();
+        // Use a maximum of 99% of the available memory as a safe-guard.
+        const auto wgroup_size = std::min(
+            {Kokkos::bit_ceil(size),
+             static_cast<size_t>(max / multiple) * multiple,
+             Kokkos::bit_floor(static_cast<size_t>(max_local_memory * .99) /
+                               (sizeof(value_type) * value_count))});
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:427]: Please comment about this (in code)
Are you are trying to get a power of two?

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp [hunk 2]
```diff
@@ -404,42 +403,60 @@ class ParallelReduce<CombinedFunctorReducerType, Kokkos::RangePolicy<Traits...>,
         auto multiple = kernel.get_info<sycl::info::kernel_device_specific::
                                             preferred_work_group_size_multiple>(
             q.get_device());
+        // FIXME_SYCL The code below queries the kernel for the maximum subgroup
+        // size but it turns out that this is not accurate and choosing a larger
+        // subgroup size gives better peformance (and is what the SYCL compiler
+        // does for sycl::reductions).
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:409]: Yes.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp [hunk 3]
```diff
@@ -300,7 +299,7 @@ class ParallelReduce<CombinedFunctorReducerType, Kokkos::RangePolicy<Traits...>,
               const auto upper_bound = std::min<index_type>(
                   global_id + values_per_thread * wgroup_size, size);
 
-              if constexpr (ReducerType::static_value_size() == 0) {
+              if constexpr (!use_shuffle_based_algorithm<ReducerType>) {
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:302]: Added.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp]: 9ac49e9

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp [hunk 4]
```diff
@@ -404,42 +403,65 @@ class ParallelReduce<CombinedFunctorReducerType, Kokkos::RangePolicy<Traits...>,
...
-        const size_t wgroup_size =
-            static_cast<size_t>(max / multiple) * multiple;
+        auto max_local_memory =
+            q.get_device().get_info<sycl::info::device::local_mem_size>();
+        // The workgroup size is computed as the minimum of
+        // - the smallest power of two not less than the total work size
+        // - the largest power of two not exceeding the largest multiple of the
+        //   recommended workgroup size not exceeding the maximum workgroup size
+        // - the largest power of two such that we don't use more than 99% (as a
+        //   safe-gurad) of the available local memory.
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:427]:
```suggestion
        //   safe-guard) of the available local memory.
```

**Review by dalg24 — APPROVED** [review]

**Comment by masterleinad** [comment]
SYCL CI is passing.

**Pull request merged**